### DISCUSSION TITLE
Article Filter: How can I remove Articles that are not in my whitelist?

### DISCUSSION THREAD

**Discussion by Crehet** [martinrotter/rssguard#987] [Q&A] [answered]
I've got a feed for [archlinux pkgs](https://archlinux.org/feeds/packages/x86_64/), and I have a lot of Articles that are just sitting there, and I want to remove them while also rejecting blacklisted keywords from getting fetched.

```js
var whitelist = [
  'firefox'
];

var blacklist = [
  'firefox-developer-edition'
];

function filterMessage() {
  if (whitelist.some(i => msg.title.indexOf(i) != -1)) {
   if (blacklist.some(i => msg.title.indexOf(i) != -1)) {
    return MessageObject.Ignore;
   }
   else {
   return MessageObject.Accept;
  }
 }
}
```

**Comment by Mitecon** [comment, marked as answer]
It's helpful to think of it in terms of:
```
I want 'this' AND NOT 'that' ELSE ignore everything else.
```
So you want to allow 'firefox' but not 'firefox-developer-edition' and ignore everything else.

However, you've got Accept and Ignore the wrong way round:
```
    return MessageObject.Ignore;
   }
   else {
   return MessageObject.Accept;
```
This is because you want to run actions on the conditions and then ignore everything else.

So we swap those:
```
    return MessageObject.Accept;
   }
   else {
   return MessageObject.Ignore;
```
Now onto the IF statement:
```
  if (whitelist.some(i => msg.title.indexOf(i) != -1)) {
   if (blacklist.some(i => msg.title.indexOf(i) != -1)) {
```
Remember, you're looking for:
```
I want 'this' AND NOT 'that' ELSE ignore everything else.
```
Firstly, what I would highly recommend is to convert everything into lower case so you don't have to account for upper and lower case characters:
```
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) {
  if (blacklist.some(i => msg.title.toLowerCase().indexOf(i) != -1)) {
```
You have two if statements there when you only need one:
```
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) &&
  (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))))
```
You need 'this' AND NOT 'that', so you use a double ampersand, '&&',  to join them together. Then the extra curly brace '{' is not needed.

To make something a NOT condition, notice how at the beginning of this line:
```
  (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))))
```
I've added '(!' meaning the exclamation mark is saying NOT. You need to excapsulate the NOT part inside a parenthesis, meaning you need to need another at the end to enclose it.

You also had an extra curly brace right at the end which is not needed.

So the final working filter would look like:
```
var whitelist = [
  'firefox'
];

var blacklist = [
  'firefox-developer-edition'
];

function filterMessage() {
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) &&
  (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))))
  {
    return MessageObject.Accept;
   }
   else {
   return MessageObject.Ignore;
  }
}
```
It sounds like you also have articles in your database already. If so and you would like to delete all of those, you can change:
```
   return MessageObject.Ignore;
```
to:
```
   return MessageObject.Purge;
```
However, I would highly suggest that you do plenty of testing using the 'Test' button, rather than the 'Process checked feeds' button first and make sure it'll definitely only delete what you expect it to.

Changing your filter to use Purge will not delete articles already in the database - you'd have to check the feed and click 'Process checked feeds' to do that.

If you leave that line as Purge, then any new articles that would be filtered out will simply be ignored and won't be added to the database.

**Reply by Crehet** [reply]
Cool. This worked just fine for me. Newbies looking for these sort of filters will appreciate this as well.

**Reply by Crehet** [reply]
By the way, the '(!' part saying 'NOT', does it mean it 's similar as `return MessageObject.Ignore;`?

I've got it my filter like this to ignore blacklisted keywords and purge every article of the feed from the DB that _is_ part of the blacklisted keywords and _is not_ in the whitelist:

```js
var whitelist = [
  'firefox'
];

var blacklist = [
  'firefox-developer-edition'
];

function filterMessage() {
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) &&
  (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))))
  {
    return MessageObject.Accept;
   }
   else {
   return MessageObject.Purge;
  }
}
```

**Comment by Mitecon** [comment]
This is a variable (or set of variables):
```
var whitelist = [
  'firefox'
];
```
This determines whether or not the articles match anything in the variables:
```
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))
```
It's saying 'IF the articles contain SOME (meaning at least one of) the variables in the WHITELIST, then this condition is TRUE or else it's FALSE'.

The '(!' part is to negate the above.

You want to keep articles containing:
```
'firefox' AND NOT 'firefox-developer-edition'
```
So you have a function in your filter (note the curly braces which opens to encapsulate the function and which is closed later on):
```
function filterMessage() {
  if ((whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) &&
  (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))))
  {
```
These conditions will evaluate to either TRUE or FALSE.

The above translates as:
```
IF the articles DO contain SOME (meaning at least one of) the variables in the WHITELIST then this condition is TRUE
        AND
   the articles DO NOT contain SOME (meaning at least one of) the variables in the BLACKLIST then this condition is TRUE
        then perform one of the the following actions
```
The actions being to either Accept the articles or Ignore or Purge them.
```
    return MessageObject.Accept;
   }
   else {
   return MessageObject.Purge;
```
If both conditions are TRUE, then the article will be Accepted, else they will be Purged/Ignored.

In the simplest terms:
```
IF  condition1 is TRUE (the variable in WHITELIST exists)
        AND
    condition2 is TRUE (the variable in BLACKLIST does not exist)
THEN
    Accept the message
ELSE
    Purge the message
```
The conditions are separate from the actions so there is no similarity there.
